---
#MIT License
#
#Copyright (c) 2021 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "gviz_practice"
output: html_document
---


```{r}

library(Gviz)
library(data.table)
library(ggplot2)
options(ucscChromosomeNames=FALSE)


```


```{r}
## gff file of gene location

df_gff <- data.frame(fread('AvrPita3.gff3'))

colnames(df_gff) <- c('chromosome',
                      'source',
                      'feature',
                      'start',
                      'end',
                      'score',
                      'strand',
                      'other_score',
                      'tag')

df_gff <- df_gff[df_gff$feature != 'gene' & df_gff$feature != 'mRNA' & df_gff$feature != 'CDS',]

## add mariner tn location, all values except scaffold, start and end are place holders
df_te <- df_gff
df_te[nrow(df_te) + 1,] = c('MQOP01000016.1',	'AUGUSTUS',	'exon',	370854, 372741,	0,	'-',	'.',	'ID=gene_0.t1.e1;Parent=gene_0.t1')
df_te <- df_te[df_te$score == 0,]
df_te$start <- as.numeric(df_te$start)
df_te$end <- as.numeric(df_te$end)



# called eccs in region
df_eccs <- data.frame(fread('AvrPita3.G3_1A.eccs.bed'))

colnames(df_eccs) <- c('chromosome',
                      'start',
                      'end')
df_eccs <- df_eccs[,c(1,2,3)]

## define sanger track based off sanger sequence analysis
df_sanger <- data.frame(matrix(ncol = 4, nrow = 0))
df_sanger[1,] <- c('MQOP01000016.1', 366830, 366830+118, '+')
df_sanger[2,] <- c('MQOP01000016.1', 373149-42, 373149, '+')
colnames(df_sanger) <- c('chromosome',
                      'start',
                      'end',
                      'strand')
df_sanger$start <- as.numeric(df_sanger$start)
df_sanger$end <- as.numeric(df_sanger$end)


```


```{r}


# genome/chromosome track
genome_track <- GenomeAxisTrack()

## make tracks

# define gene region
grtrack <- GeneRegionTrack(df_gff, name = " ", gene='gene_11229', transcript='gene_11229.t1',
                            symbol = NULL, fontsize.group=30, 
                           col='#000000', shape = "arrow", fill='#006400', fontcolor.group="#000000",
                           fontcolor.title='#000000', rotation.title=0, fontsize=20)

tetrack <- GeneRegionTrack(df_te, name = " ", gene='gene_0', transcript='gene_0.t1',
                            symbol = NULL, fontsize.group=30, 
                           col='#000000', shape = "box", fill='orange', fontcolor.group="#000000",
                           fontcolor.title='#000000', rotation.title=0, fontsize=20)

grtrack <- OverlayTrack(trackList=list(grtrack,tetrack))

# define ecc track
ecc_track <- AnnotationTrack(df_eccs, name = ' ', col = '#000000', fill='#000000', min.height=6, max.height=6, # set size of track
                              fontcolor.title='#000000', rotation.title=0, fontsize=20)
  
# define sanger track
sanger_seq <- AnnotationTrack(df_sanger, name = ' ', col = '#000000', fill='#FF0000', min.height=6, max.height=6, # set size of track
                              fontcolor.title='#000000', rotation.title=0,strand=c('+','+'), shape='arrow', fontsize=20)

#define total reads track
total_reads <- AlignmentsTrack(
  'AvrPita3.G3_1A.totalreads.sorted.bam', isPaired = TRUE,
           chromosome = 'MQOP01000016.1', type = "coverage", name = ' ',
  coverageHeight=1.4, col.coverage='#000000', showAxis=FALSE, col.axis='transparent',
   fontcolor.title='#000000', rotation.title=0, fontsize=20)


# define junction split read tracks
split_reads <- AlignmentsTrack(
  'junction_splitreads.AvrPita3.G3_1A.sorted.bam', isPaired = TRUE,
           chromosome = 'MQOP01000016.1', type = "pileup", name = ' ', lty.mates = 2, fill.reads='#F8766D', min.height=5, max.height=5, # this sets size of arrows
  col.mates='#000000', col.reads='#000000',
   fontcolor.title='#000000', rotation.title=0, fontsize=20)

# define opposite facing read track
discordant_reads <- AlignmentsTrack(
  'outwardfacing.AvrPita3.G3_1A.sorted.bam', isPaired = TRUE,
           chromosome = 'MQOP01000016.1', type = "pileup", name = ' ', fill.reads='#00BFC4', min.height=5, max.height=5, # this sets size of arrows
            col.mates='#000000', col.reads='#000000',
   fontcolor.title='#000000', rotation.title=0, fontsize=20)

# define pacbio split reads track
pacbio_splitreads <- AlignmentsTrack(
  'AvrPita3.G3_1A_pacbio_splitreads.sorted.bam', isPaired = TRUE, # this is necessary for overlap to happen
           chromosome = 'MQOP01000016.1', type = "pileup", name = ' ', lty.mates = 2,
           col.reads='#000000', min.height=5, max.height=5, # make them pretty big
   fontcolor.title='#000000', rotation.title=0, fontsize=20)




pdf("avrpita3_on_eccdnas_G3_1A.pdf", width=13, height=14) # need to make it double size and scale down otherwise the arrows arent properly drawn...
plotTracks(list(genome_track,
                grtrack,
                ecc_track,
                sanger_seq,
                total_reads,
                split_reads,
                discordant_reads,
                pacbio_splitreads), sizes=c(0.15, #genome
                                            0.05, #genes
                                            0.25, #ecc
                                            0.05, # sanger
                                            0.125, # read cov
                                            1.0, # split reads
                                            1.0, # opposite facing reads
                                            0.15), # pacbio
           from = 368546-2500, to = 370280+3000, showTitle=TRUE, showAxis=FALSE) # define region
dev.off()


```

