---
#MIT License
#
#Copyright (c) 2021 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "gene_portions_enrichment_plot"
output: html_document
---


```{r}
library(data.table)
library(ggplot2)
library(tidyr)

# read in table and name columns

df <- data.frame(fread('gene_portions_full_table'))

colnames(df) <- c("sample", "gene_portion",'observed_micro', 'expected_micro', 'observed_large', 'expected_large')
```

```{r}

# separate out sample metadata into separate columns

df$sample_full <- df$sample

df <- separate(data=df, col=sample, into=c('study', 'organism', 'tissue', 'genotype','treatment'), sep ="=")


```


```{r}

# log2 of observed over expected gives log2 fold enrichment

df$log2_foldchange_micro <- log2(df$observed_micro/df$expected_micro)

df$log2_foldchange_large <- log2(df$observed_large/df$expected_large)

df$gene_portion <- factor(df$gene_portion, levels=c('exons', 'introns', 'upstream', 'downstream', 
                                                    'five_prime_utr', 'three_prime_utr', 'cpg_islands', 'genic', 'intergenic'))

```


```{r}

# just making full labels

df$labels <- c(
rep('S. cerevisiae deletion collection \n(Moller et al. 2015)',9),
rep('S. cerevisiae deletion collection, zeocin treatment \n(Moller et al. 2015)',9),
rep('S. cerevesiae GAP1circle \n(Moller et al. 2015)',9),
rep('S. cerevesiae \n(Moller et al. 2015)',9),
rep('A. thaliana flower epi12 mutant \n(Lanciano et al 2017)',9),
rep('A. thaliana flower \n(Lanciano et al 2017)',9),
rep('O. sativa callus \n(Lanciano et al 2017)',9),
rep('O. sativa leaf \n(Lanciano et al 2017)',9),
rep('O. sativa seed \n(Lanciano et al 2017)',9),
rep('H. sapiens leukocyte \n(Moller et al 2018)',9),
rep('H. sapiens muscle \n(Moller et al 2018)',9),
rep('A. thaliana flower \n(Wang et al 2020)',9),
rep('A. thaliana leaf \n(Wang et al 2020)',9),
rep('A. thaliana root \n(Wang et al 2020)',9),
rep('A. thaliana stem \n(Wang et al 2020)',9),
rep('O. sativa leaf (this study)',9),
rep('M. oryzae (this study)',9)
)

order <- c(
'S. cerevesiae \n(Moller et al. 2015)',
'S. cerevesiae GAP1circle \n(Moller et al. 2015)',
'S. cerevisiae deletion collection \n(Moller et al. 2015)',
'S. cerevisiae deletion collection, zeocin treatment \n(Moller et al. 2015)',
'H. sapiens leukocyte \n(Moller et al 2018)',
'H. sapiens muscle \n(Moller et al 2018)',
'A. thaliana leaf \n(Wang et al 2020)',
'A. thaliana root \n(Wang et al 2020)',
'A. thaliana stem \n(Wang et al 2020)',
'A. thaliana flower \n(Wang et al 2020)',
'A. thaliana flower \n(Lanciano et al 2017)',
'A. thaliana flower epi12 mutant \n(Lanciano et al 2017)',
'O. sativa callus \n(Lanciano et al 2017)',
'O. sativa seed \n(Lanciano et al 2017)',
'O. sativa leaf \n(Lanciano et al 2017)',
'O. sativa leaf (this study)',
'M. oryzae (this study)'
)

df$labels <- factor(df$labels, levels=order)



```


```{r}

# plot for microdnas across all organisms and sample types

p <- ggplot(data=df, aes(x=gene_portion, y=log2_foldchange_micro, fill=labels))+
  geom_bar(stat='identity', position=position_dodge()) + 
  geom_vline(xintercept = seq(1.5, 9, by = 1), color="black", size=.5, alpha=1)+
  coord_flip()+
  theme_classic()+
  ylab('Log2 fold enrichment')+
  guides(fill = guide_legend(reverse=TRUE))+ # reverse legend order
  scale_x_discrete(labels=c('Exons', 'Introns', 'Upstream', 'Downstream',
                              "5'UTR", "3'UTR",'CpG islands',
                              'Genic', 'Intergrenic'))+ # rename x-axis
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.text=element_text(size=6),
        axis.title.y=element_blank())


p

ggsave('gene_portions_plot_full_micro.pdf', plot=p, width=6.5, height=6.5, device = 'pdf')


```


```{r}
# plot for large eccdnas across all organisms and sample types

p <- ggplot(data=df, aes(x=gene_portion, y=log2_foldchange_large, fill=labels))+
  geom_bar(stat='identity', position=position_dodge()) + 
  geom_vline(xintercept = seq(1.5, 9, by = 1), color="black", size=.5, alpha=1)+
  coord_flip()+
  theme_classic()+
  ylab('Log2 fold enrichment')+
  scale_x_discrete(labels=c('Exons', 'Introns', 'Upstream', 'Downstream',
                              "5'UTR", "3'UTR",'CpG islands',
                              'Genic', 'Intergrenic'))+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.text=element_text(size=6),
        axis.title.y=element_blank())


p

ggsave('gene_portions_plot_full_large.pdf', plot=p, width=6.5, height=6.5, device = 'pdf')
```


```{r}

# subset to representative samples

df_human_muscle <- df[df$organism=="Human" & df$tissue!="Leukocyte",]

df_ara_flower <- df[df$organism=="Arabidopsis" & df$tissue=="Flower" & df$study=="2020_Wang",]

df_rice_leaf <- df[df$organism=="Rice" & df$tissue=="Leaf" & df$study=="2017_Lanciano",]

df_riceblast <- df[df$organism=="Rice_blast",]

df_representative <- rbind(df_human_muscle, df_ara_flower, df_rice_leaf, df_riceblast)


  
```




```{r}

# representative samples for microdnas

p <- ggplot(data=df_representative, aes(x=gene_portion, y=log2_foldchange_micro, fill=labels))+
  geom_bar(stat='identity', position=position_dodge()) + 
  geom_vline(xintercept = seq(1.5, 9, by = 1), color="black", size=.5, alpha=1)+
  coord_flip()+
  theme_classic()+
  ylab('Log2 fold enrichment')+
  scale_x_discrete(labels=c('Exons', 'Introns', 'Upstream', 'Downstream',
                              "5'UTR", "3'UTR",'CpG islands',
                              'Genic', 'Intergrenic'))+
      guides(fill=guide_legend(nrow=2,byrow=TRUE,keywidth=0.1, keyheight=0.1, default.unit='inch'))+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.text=element_text(size=6),
        axis.title.y=element_blank(), legend.position='bottom')



p

ggsave('gene_portions_plot_representative_micro.pdf', plot=p, width=3.25, height=3.25, device = 'pdf')


```

```{r}

## only looking at rice blast now
# prepare data for plotting

expected_large_vector <- df$expected_large[df$organism=="Rice_blast"]

expected_micro_vector <- df$expected_micro[df$organism=="Rice_blast"]

# read in moryzae data per sample

df_moryzae_large <- data.frame(fread('moryzae.observed_averages_large_all_samples'))
df_moryzae_large <- subset(df_moryzae_large, select = -c(V10))
colnames(df_moryzae_large) <- c('gene_portion', 'G3_1A', 'G3_1B', 'G3_1C',
                                'G3_2A', 'G3_2B', 'G3_2C',
                                'G3_3A', 'G3_3B')

rownames(df_moryzae_large) <- df_moryzae_large$gene_portion
df_moryzae_large <- subset(df_moryzae_large, select = -c(gene_portion))

# get log fold change as before
for (i in seq(from=1, to=length(expected_large_vector))) {
  number = expected_large_vector[[i]]
  vector = df_moryzae_large[i,1:8]
  transformed <- log2(vector/number)
  df_moryzae_large[i,1:8] <- transformed
}

df_moryzae_large_stacked <- stack(df_moryzae_large)

colnames(df_moryzae_large_stacked) <- c('fold_change','sample')

df_moryzae_large_stacked$gene_portion <- rep(rownames(df_moryzae_large), 8)
df_moryzae_large_stacked$ecc <- rep('large', nrow(df_moryzae_large_stacked))

# read in moryzae data per sample

df_moryzae_micro <- data.frame(fread('moryzae.observed_averages_micro_all_samples'))
df_moryzae_micro <- subset(df_moryzae_micro, select = -c(V10))
colnames(df_moryzae_micro) <- c('gene_portion', 'G3_1A', 'G3_1B', 'G3_1C',
                                'G3_2A', 'G3_2B', 'G3_2C',
                                'G3_3A', 'G3_3B')

rownames(df_moryzae_micro) <- df_moryzae_micro$gene_portion
df_moryzae_micro <- subset(df_moryzae_micro, select = -c(gene_portion))

# get log fold change as before
for (i in seq(from=1, to=length(expected_micro_vector))) {
  number = expected_micro_vector[[i]]
  vector = df_moryzae_micro[i,1:8]
  transformed <- log2(vector/number)
  df_moryzae_micro[i,1:8] <- transformed
}

df_moryzae_micro_stacked <- stack(df_moryzae_micro)

colnames(df_moryzae_micro_stacked) <- c('fold_change','sample')

df_moryzae_micro_stacked$gene_portion <- rep(rownames(df_moryzae_micro), 8)
df_moryzae_micro_stacked$ecc <- rep('micro', nrow(df_moryzae_micro_stacked))


df_moryzae <- rbind(df_moryzae_large_stacked, df_moryzae_micro_stacked)

df_moryzae$gene_portion <- factor(df_moryzae$gene_portion, levels=c('exons', 'introns', 'upstream', 'downstream', 
                                                    'five_prime_utr', 'three_prime_utr', 'cpg_islands', 'genic', 'intergenic'))

```

```{r}
troubleshooting

df_moryzae_large

log2(rowMeans(df_moryzae_large)/expected_large_vector)
```



```{r}

# enrichment for large and microdnas in magnaporthe only

p <- ggplot(data=df_moryzae, aes(x=gene_portion, y=fold_change, color=ecc))+
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(position=position_jitterdodge(), aes(color=ecc), size =0.5)+
  coord_flip()+
  theme_classic()+
  geom_hline(yintercept = 0)+
  ylab('Log2 fold enrichment')+
  scale_x_discrete(labels=c('Exons', 'Introns', 'Upstream', 'Downstream',
                              "5'UTR", "3'UTR",'CpG islands',
                              'Genic', 'Intergrenic'))+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.text=element_text(size=6),
        axis.title.y=element_blank(), legend.position='bottom')

p

ggsave('gene_portions_plot_large_v_micro.pdf', plot=p, width=6.5, height=3.25, device = 'pdf')

```

