---
#MIT License
#
#Copyright (c) 2021 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "minichromosome_genes_on_eccdnas_analysis"
output: html_document
---


```{r}
library(data.table)
library(ggplot2)

# read in encompassing split reads per gene

df <- data.frame(fread('G3.normalized.splitreadspergene'))


# read in list of genes on minichromosome
minichrom_genes <- read.table('all_genes_on_minichromosome', header=FALSE, sep='\t')

# tag minichromosome genes
minichrom = function(x, output) {
  id = x[[1]]
  if (is.element(id, minichrom_genes$V1)){
    return(TRUE)
  } else{
    return(FALSE)
  }
}

df <- df[df$V3 != 0,]

df$V3 <- log10(df$V3)

df$V4 <- apply(df, 1, minichrom)


colnames(df) <- c('gene','rpkm','srs', 'minichrom')
```

```{r}
# simple box plot of encompassing split reads between genes on minichromosomes and other genes

labs <- c("Other genes", "Genes on \nmini-chromosomes")

p <- ggplot(df, aes(x=minichrom, y=srs)) + geom_boxplot() + theme_classic() + ylab('Log 10 encompassing split \nread count') + xlab('') + scale_x_discrete(labels= labs)+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_text(size=8), legend.position = 'bottom', legend.text=element_text(size=6))


p

ggsave("minichromosome_genes_on_eccdnas_boxplot.pdf", plot = p, width = 2, height = 2)

```

```{r}

## permutation test for signifiance between both groups
observed <- median(df$srs[df$minichrom == TRUE ]) - median(df$srs[df$minichrom == FALSE ])

# build null distribution
permutation = replicate(10000, {
  sample_small <- df[sample(nrow(df), nrow(df[df$minichrom == TRUE,]) , replace = FALSE), ]# one sample the size of minichrom genes
  sample_large <- df[! rownames(df) %in% rownames(sample_small), ]# one sample with the rest
  median(sample_small$srs)-median(sample_large$srs)# calculate media between the two
})

p <- ggplot() + aes(permutation)+ geom_histogram(binwidth=0.01, colour="black", fill="white") + xlim(-1,1) + geom_vline(xintercept=observed, size=1.5)

p

## pvalue is number of permutations that are bigger than observed

p_value <- mean(permutation > observed)
p_value
```

