---
#MIT License
#
#Copyright (c) 2021 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "coverage_plot_ggplot_never_v_common"
output: html_document
---

```{r}
library(data.table)
library(ggplot2)
library(RColorBrewer)
library(ggpubr)
library(ggrepel)

```

```{r}

# genes
df <- data.frame(fread('guy11_fungap_out_12_28_20.bed'))

# encompassing split reads per gene
df_splitreads <- data.frame(fread('G3.normalized.splitreadspergene'))

colnames(df) <- c('CHROMOSOME', 'start', 'end', 'gene')

## a lot of work to draw one dot at the middle of each gene

df$CHROMOSOME <- as.numeric(substring(df$CHROMOSOME, 11, 12))
                          
df$middle <- df$start + (df$end - df$start) * 0.5

df$splitreads <- df_splitreads$V2

df$status <- NA

df$status[df$splitreads == 0] <- 'eccDNA-absent'

# call common and never found genes based off previously identified cutoffs

common_genes <- data.frame(fread('G3.common.genes', header=FALSE))

common_genes <- common_genes$V1

df$status[df$gene %in% common_genes] <- 'eccDNA-associated'

df$status[is.na(df$status)] <- 'Other genes'

```

```{r}

# scaffold and scaffold lengths
df_chromsizes <- data.frame(fread('guy11_genome_baoetal2017.chromsizes'))

colnames(df_chromsizes) <- c('chrom', 'size')
df_chromsizes$chrom <- as.numeric(substring(df_chromsizes$chrom, 11, 12))

## count base pairs throughout the gneome according to which scaffold they are on, this is to plot the genome in a way that makes sense

v <- numeric()

for (scaffold in 1:56){
  v[scaffold] <- df_chromsizes$size[df_chromsizes$chrom == scaffold]
}

x <- numeric()

for (scaffold in 1:56){
  x[scaffold] <- sum(v[1:scaffold]) - v[scaffold]
}


f <- function(g) {
 base <- as.numeric(g[5])
 scaffold <- as.numeric(g[1])
 base_total <- base + x[scaffold]
 return(base_total)
}

df$BASE_TOTAL <- apply(df, 1, f)

df$status <- factor(df$status, levels=c('eccDNA-absent', 'Other genes', 'eccDNA-associated'))

# so we can take the log of zeroes
df$splitreads[df$splitreads == 0] <- 0.001

df$splitreads <- log10(df$splitreads)

```

```{r}
# add known genes name

known_genes_files = 'guy11_70_15_gene_names.txt'
known_genes_table = read.table(known_genes_files, header=FALSE, sep='\t')

grab_splitreads <- function(x) {
  return(df$splitreads[df$gene == x[2]])
}
grab_base <- function(x) {
  return(df$BASE_TOTAL[df$gene == x[2]])
}

known_genes_table$V3 <- apply(known_genes_table, 1, grab_splitreads)
known_genes_table$V4 <- apply(known_genes_table, 1, grab_base)

  
colnames(known_genes_table) <- c('name_70_15', 'gene', 'splitreads', 'BASE_TOTAL')


known_genes_table$name_70_15 <- as.character(known_genes_table$name_70_15)

# filter to only named genes
known_genes_table <- known_genes_table[known_genes_table$name_70_15 != 'unknown',]

known_genes_table <- known_genes_table[! startsWith(known_genes_table$name_70_15,'MGG_'),]

known_genes_table <- known_genes_table[! startsWith(known_genes_table$name_70_15,'MGCH'),]

# sort genes by absent or associated

known_genes_table_associated <- known_genes_table[known_genes_table$gene %in% common_genes,]

known_genes_table_absent <- known_genes_table[known_genes_table$splitreads == min(known_genes_table$splitreads),]

```

```{r}

## coverage plot for first few scaffolds

start_subset <- 1

end_subset <- 5

end_subset_first_few <- 5

count <- end_subset - start_subset + 1

ticks <- numeric()

for (scaffold in start_subset:end_subset){
  ticks <- c(ticks, v[scaffold]*0.5 + (sum(v[start_subset:scaffold]) - v[scaffold]))
  ticks <- c(ticks, v[scaffold]*1 + (sum(v[start_subset:scaffold]) - v[scaffold]))
}

labs <- rep(start_subset:end_subset, each=2)


df_subset <- df[df$CHROMOSOME >= start_subset & df$CHROMOSOME <= end_subset,]

p1 <- ggplot() + 
  geom_point(data = df_subset[df_subset$status=="Other genes",], aes(x=BASE_TOTAL, y=splitreads, color='Other genes'),size=0.1)+
  geom_point(data = df_subset[df_subset$status=="eccDNA-absent",], aes(x=BASE_TOTAL, y=splitreads, color='eccDNA-absent'),size=0.1) +
  geom_point(data = df_subset[df_subset$status=="eccDNA-associated",], aes(x=BASE_TOTAL, y=splitreads, color='eccDNA-associated'),size=0.1) +
      geom_text(data = known_genes_table_associated, aes(x=BASE_TOTAL, y=splitreads, label=name_70_15), size =2, fontface="italic") +
      geom_text_repel(data = known_genes_table_absent, aes(x=BASE_TOTAL, y=splitreads, label=name_70_15), size =2, min.segment.length = 0, max.overlaps = 100, segment.linetype = 2, force = 0.075,fontface="italic") + # repel is great here to make the labels not run into each other much
  scale_x_continuous(labels=labs, breaks = ticks,
                     limits = c((sum(v[1:start_subset]) - v[start_subset]),
                           sum(v[1:start_subset]) - v[start_subset] + sum(v[1:end_subset_first_few]))) + # setting ticks is really hard here 
  theme_classic() + xlab('Scaffold') +
  ylim(0, max(df$splitreads)) +
  theme(axis.text.x = element_text(colour=c(rep(c('black','transparent'),count))))+ # need to draw invisible labels so the plot looks pretty
  theme(axis.ticks.x = element_line(colour=c(rep(c('transparent','black'),count))))+ # need to draw invisible ticks so the plot looks pretty
  scale_y_continuous(breaks=-3:1, labels=c("0","0.01", "0.1", "1", "10"), limits = c(-3,1)) +
  ylab('Normalized encompassing \nsplit read count')+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.position = 'bottom', legend.text=element_text(size=6)) + guides(colour = guide_legend(override.aes = list(size=2))) +
  theme(axis.line.y = element_blank()) +
  annotate(geom = "segment", x = -Inf, xend = -Inf, y = -Inf, yend = Inf) +
  annotate(geom = "segment", x = -Inf, xend = -Inf, y =  -2.75, yend = -2.5, color = "white", size = 1) + # chop up y axis which is not continuous because of zeroes
  scale_colour_manual(breaks = c("eccDNA-absent", "Other genes", "eccDNA-associated"),
                      values = c("eccDNA-absent" = '#03254c', "Other genes" = "grey", "eccDNA-associated" = "#E6AB02"))


ggsave('test.tiff', plot=p1, width=6.5, height=1.75, device = 'tiff', dpi = 1000)

```

```{r}

## coverage plot for first remaining scaffolds


start_subset <- 6

end_subset <- 56

count <- end_subset - start_subset + 1

ticks <- numeric()

for (scaffold in start_subset:end_subset){
  if (scaffold <= 15) {
    ticks <- c(ticks, v[scaffold]*0.5 + (sum(v[start_subset:scaffold]) - v[scaffold]))
    } else if (scaffold == 20) {
      ticks <- c(ticks, v[scaffold]*0.5 + (sum(v[start_subset:scaffold]) - v[scaffold]))
    } 
  ticks <- c(ticks, v[scaffold]*1 + (sum(v[start_subset:scaffold]) - v[scaffold]))
}

ticks <- ticks + sum(v[1:end_subset_first_few])

labs <- character()

for (scaffold in start_subset:end_subset){
  if (scaffold <= 15) {
    labs <- c(labs, scaffold)
    } else if (scaffold == 20) {
      labs <- c(labs,scaffold)
    } 
  labs <- c(labs, scaffold)
}

#tick colors to make it easier
# especially since we arent labeling every label because of the extra, closely clusetered scaffolds

tick_colors <- character()
label_colors <- character()

for (scaffold in start_subset:end_subset){
  if (scaffold <= 15) {
    tick_colors <- c(tick_colors, 'transparent', 'black')
    } else if (scaffold == 20) {
      tick_colors <- c(tick_colors, 'transparent', 'black')
    } else {
      tick_colors <- c(tick_colors, 'black')
    }
  if (scaffold <= 15) {
    label_colors <- c(label_colors, 'black', 'transparent')
    } else if (scaffold == 20) {
      label_colors <- c(label_colors, 'black', 'transparent')
    } else {
      label_colors <- c(label_colors, 'transparent')
    }
}



df_subset <- df[df$CHROMOSOME >= start_subset & df$CHROMOSOME <= end_subset,]

p2 <- ggplot() +
  geom_point(data = df_subset[df_subset$status=="Other genes",], aes(x=BASE_TOTAL, y=splitreads, color='Other genes'),size=0.1)+
  geom_point(data = df_subset[df_subset$status=="eccDNA-absent",], aes(x=BASE_TOTAL, y=splitreads, color='eccDNA-absent'),size=0.1) +
  geom_point(data = df_subset[df_subset$status=="eccDNA-associated",], aes(x=BASE_TOTAL, y=splitreads, color='eccDNA-associated'),size=0.1) +
        geom_text(data = known_genes_table_associated, aes(x=BASE_TOTAL, y=splitreads, label=name_70_15), size =2, fontface="italic") +
      geom_text_repel(data = known_genes_table_absent, aes(x=BASE_TOTAL, y=splitreads, label=name_70_15), size =2, min.segment.length = 0, max.overlaps = 100, segment.linetype = 2, force = 0.075,fontface="italic") +# repel is great here to make the labels not run into each other much
  scale_x_continuous(labels=labs, breaks = ticks,
                     limits = c((sum(v[1:start_subset]) - v[start_subset]),
                           sum(v[1:start_subset]) - v[start_subset] + sum(v[1:end_subset_first_few]))) +  # setting ticks is really hard here 
  theme_classic() + xlab('Scaffold') +
  ylim(0, max(df$splitreads)) +
  theme(axis.text.x = element_text(colour=label_colors))+# need to draw invisible labels so the plot looks pretty
  theme(axis.ticks.x = element_line(colour=tick_colors))+# need to draw invisible ticks so the plot looks pretty
  scale_y_continuous(breaks=-3:1, labels=c("0","0.01", "0.1", "1", "10"), limits = c(-3,1)) +
  ylab('Normalized encompassing \nsplit read count')+
  theme(axis.text=element_text(size=6),axis.title=element_text(size=8) , legend.title=element_blank(), legend.position = 'bottom', legend.text=element_text(size=6))+ guides(colour = guide_legend(override.aes = list(size=2)))+
  theme(axis.line.y = element_blank()) +
  annotate(geom = "segment", x = -Inf, xend = -Inf, y = -Inf, yend = Inf) +
  annotate(geom = "segment", x = -Inf, xend = -Inf, y =  -2.75, yend = -2.5, color = "white", size = 1)+# chop up y axis which is not continuous because of zeroes
  scale_colour_manual(breaks = c("eccDNA-absent", "Other genes", "eccDNA-associated"),
                      values = c("eccDNA-absent" = '#03254c', "Other genes" = "grey", "eccDNA-associated" = "#E6AB02"))


# should be 5000 for high res
ggsave('test2.tiff', plot=p2, width=6.5, height=1.75, device = 'tiff', dpi = 1000)

```

```{r}

## ggarrange so they look well aligned
both <- ggarrange(p1,p2, ncol=1, nrow=2, common.legend = TRUE, legend='bottom')

ggsave('test3.tiff', plot=both, width=6.5, height=3.5, device = 'tiff', dpi = 1000)

ggsave('never_v_common_coverage_plot.pdf', plot=both, width=6.5, height=3.5, device = 'pdf')

```

